src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -392,11 +398,18 @@ fn rustdoc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
     }
 
     let name = unit.pkg.name().to_string();
-    let desc = rustdoc.to_string();
+    let build_state = cx.build_state.clone();
+    let key = (unit.pkg.package_id().clone(), unit.kind);
     let exec_engine = cx.exec_engine.clone();
 
     Ok(Work::new(move |desc_tx| {
-        desc_tx.send(desc).unwrap();
+        let outputs = build_state.outputs.lock().unwrap();
```

Comment:
This'll actually want to drop the lock ASAP instead of holding it over the entire call to `rustdoc`